perf(core): Make `PlatformLocation` tree-shakable

mhevery opened the pull request:
Convert `PlatformLocation` into a tree-shakable provider

alfaproject commented on packages/compiler-cli/test/compliance/r3_compiler_compliance_spec.ts:
```diff
@@ -2785,7 +2785,7 @@ describe('compiler compliance', () => {
               $r3$.ɵɵtext(2);
               $r3$.ɵɵelementEnd();
-              $r3$.ɵɵelementStart(3, "ul");
+              $r3$.��ɵelementStart(3, "ul");
```
👀

mhevery replied:
Not ready for review yet. :-)

alxhub commented on packages/core/src/di/scope.ts:
```diff
@@ -14,5 +14,4 @@ import {InjectionToken} from './injection_token';
  */
-export const APP_ROOT = new InjectionToken<boolean>(
-    'The presence of this token marks an injector as being the root injector.');
+export const INJECTOR_SCOPE = new InjectionToken<'root'|'platform'|null>('Set Injector scope.');
```
Just FYI, I believe there may be some projects in g3 which make use of this private API.

mhevery replied:
thanks for the heads up

alxhub commented on packages/core/src/di/r3_injector.ts:
```diff
@@ -128,7 +130,8 @@ export class R3Injector {
     // any injectable scoped to APP_ROOT_SCOPE.
-    this.isRootInjector = this.records.has(APP_ROOT);
+    const record = this.records.get(INJECTOR_SCOPE);
+    this.scope = record ? record.value : null;
```
Use `record != null`, per @pkozlowski-opensource's perf talk about the cost of implicit conversions.

mhevery replied:
fixed

alxhub commented on packages/core/src/di/r3_injector.ts:
```diff
@@ -83,8 +83,10 @@ export function createInjector(
 export class R3Injector {
   /**
    * Map of tokens to records which contain the instances of those tokens.
+   * - `null` value implies that we don't have the record. Used by tree-shakable injectors
```
Clever! 👍

alxhub commented on packages/platform-server/src/server.ts:
```diff
@@ -27,6 +27,8 @@ function notSupported(feature: string): Error {
   throw new Error(`platform-server does not support '${feature}'.`);
 }
 
+type __retain_for_correct_d_ts_generation__ = [PlatformRef];
```
Maybe a comment here describing what happens if this isn't here?

mhevery replied:
I realized that the issue is that we rely on type inference. So I just properly typed the code and there is no need for this any more.

alxhub reviewed (approved):
Overall this LGTM.

The only thing I would suggest is this PR is one commit without much of a description, and it does a lot more than it says in the commit title. I would recommend breaking the PR into 2-3 commits (esp. one for the change to tree-shakeable provider infrastructure which allows platform providers). At the very least the commit description should document what changes were made to the overall infrastructure to allow the conversion of `PlatformLocation` to be tree-shakeable.

mhevery committed:
refactor: Move `dom_adapter.ts` to `@angular/common`

This work is needed in preparation for turning tokens into tree-shakable injectables.

mhevery committed:
feat(core): Adds DI support for `providedIn: 'platform'|'any'`

Extend the vocabulary of the `providedIn` to also include  `'platform'` and `'any'`` scope.
```
@Injectable({
  providedId: 'platform', // tree shakable injector for platform injector
})
class MyService {...}
```

mhevery committed:
perf(core): Make `PlatformLocation` tree-shakable

Convert `PlatformLocation` into a tree-shakable provider.

mhevery commented:
@alxhub Thanks for the feedback. I have broken the PR into 3 SHAs.

mhevery commented:
[presubmit](http://test/OCL:266015776:BASE:266015817:1567033668112:7cc8a7fc)
[presubmit ivy](http://test/OCL:266015776:BASE:266016098:1567033759850:ddf6c5a)

mhevery commented:
[presubmit](http://test/OCL:266015776:BASE:266026091:1567036864109:deeb8b0b)
[presubmit ivy](http://test/OCL:266015776:BASE:266026761:1567037014301:d6e6c758)

mhevery commented:
[presubmit](http://test/OCL:266015776:BASE:266058971:1567052511675:4756a895)
[presubmit ivy](http://test/OCL:266015776:BASE:266058971:1567052511675:4756a895)

mhevery commented:
[presubmit](http://test/OCL:266226030:BASE:266234225:1567115271547:8e3a6375)
[presubmit clean](http://test/OCL:266226030:BASE:266232844:1567114940361:e9030330)

mhevery commented:
[presubmit](http://test/OCL:266226030:BASE:266244387:1567118035327:e2aed72)
[presubmit clean](http://test/OCL:266226030:BASE:266246602:1567118678479:9f81aba9)

mhevery commented:
MERGE ASSISTANCE: needs cl/266062693 patch

razvanm commented:
As far as I can tell this PR broke the overriding the `Location.stripTrailingSlash` mentioned in https://github.com/angular/angular/issues/16051#issuecomment-330465196. Is that expected?

mhevery commented:
@razvanm What specifically broke. We just moved code around, there should be no changes in behavior.

razvanm commented:
In the code I'm working with I had in my root app the following code:

```
Location.stripTrailingSlash = (url: string) => {
  return url.endsWith('/') && url.length > 1 ? url + '.' : url;
};
```

This used to work fine both in dev mode and in the production code that gets compiled using the Closure compiler. After this PR the overriding stopped working (I used some `console.log()` calls to figure out that the Angular code is called instead of mine).

mhevery commented:
Right, you are using unsupported API. We don't guarantee that you can
override our API, it is not part of our public contract.

On Wed, Sep 4, 2019 at 10:19 PM Razvan Musaloiu-E. <notifications@github.com>
wrote:

> In the code I'm working with I had in my root app the following code:
>
> Location.stripTrailingSlash = (url: string) => {
>   return url.endsWith('/') && url.length > 1 ? url + '.' : url;
> };
>
> This used to work fine both in dev mode and in the production code that
> gets compiled using the Closure compiler. After this PR the overriding
> stopped working (I used some console.log() calls to figure out that the
> Angular code is called instead of mine).
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/32154?email_source=notifications&email_token=AAA3KT4B66SC4C6LI6AEYFTQICJHNA5CNFSM4IMBDUFKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD553MUQ#issuecomment-528201298>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAA3KT7HCRTL5HGH2HYWLJTQICJHNANCNFSM4IMBDUFA>
> .
>
